### ISSUE TITLE
SQLServer - Error when using a filtered Count on a group

### ISSUE THREAD

**Issue by Corey-M** [linq2db/linq2db#1078]
When using the [`Count(predicate)`][1] method to count grouped records that satisfy a filter predicate I am receiving an exception at query generation time.  I have used this technique numerous times with the default Linq2SQL query generator without any issues.

The initial query looked like this:

```
var query = 
    from u in db.Users
    group u.Active ? 1 : 0 by u.SiteID into grp
    select new
    {
        SiteID = grp.Key,
        Total = grp.Count(),
        Inactive = grp.Count(_ => _ == 0)
    };
```

This works for the default query generator but fails with Linq2DB.  Tried on several versions in my program as well as testing in LINQPad.

On enumeration (in LINQPad using the Linq2DB driver, Linq2db version 1.0.7.6) I get the following exception:

```
Exception Message:Property 'Int32 UserID' is not defined for type 'System.Int32'
Stack Trace:

at System.Linq.Expressions.Expression.Property(Expression expression, PropertyInfo property)   at System.Linq.Expressions.Expression.MakeMemberAccess(Expression expression, MemberInfo member)   at LinqToDB.Linq.Builder.ExpressionBuilder.GetParameter(Expression ex, MemberInfo member) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1872   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertObjectComparison(ExpressionType nodeType, IBuildContext leftContext, Expression left, IBuildContext rightContext, Expression right) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1798   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertCompare(IBuildContext context, ExpressionType nodeType, Expression left, Expression right) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1485   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertPredicate(IBuildContext context, Expression expression) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1311   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSearchCondition(IBuildContext context, Expression expression, List`1 conditions) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 2307   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildWhere(IBuildContext parent, IBuildContext sequence, LambdaExpression condition, Boolean checkForSubQuery, Boolean enforceHaving) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 45   at LinqToDB.Linq.Builder.WhereBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in c:\projects\linq2db\Source\Linq\Builder\WhereBuilder.cs:line 18   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo) in c:\projects\linq2db\Source\Linq\Builder\MethodCallBuilder.cs:line 22   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.cs:line 177   at LinqToDB.Linq.Builder.CountBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in c:\projects\linq2db\Source\Linq\Builder\CountBuilder.cs:line 21   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo) in c:\projects\linq2db\Source\Linq\Builder\MethodCallBuilder.cs:line 22   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.cs:line 177   at LinqToDB.Linq.Builder.ExpressionBuilder.GetSubQuery(IBuildContext context, MethodCallExpression expr) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 237   at LinqToDB.Linq.Builder.ExpressionBuilder.GetSubQueryContext(IBuildContext context, MethodCallExpression expr) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 276   at LinqToDB.Linq.Builder.ExpressionBuilder.GetSubQueryExpression(IBuildContext context, MethodCallExpression expr) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 287   at LinqToDB.Linq.Builder.ExpressionBuilder.<>c__DisplayClass57.<BuildExpression>b__55(Expression expr) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 215   at LinqToDB.Expressions.Extensions.Transform(Expression expr, Func`2 func) in c:\projects\linq2db\Source\Expressions\Extensions.cs:line 1283   at LinqToDB.Expressions.Extensions.Transform2[T](IEnumerable`1 source, Func`2 func) in c:\projects\linq2db\Source\Expressions\Extensions.cs:line 1267   at LinqToDB.Expressions.Extensions.Transform(Expression expr, Func`2 func) in c:\projects\linq2db\Source\Expressions\Extensions.cs:line 1479   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildExpression(IBuildContext context, Expression expression) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 28   at LinqToDB.Linq.Builder.SelectContext.BuildExpression(Expression expression, Int32 level) in c:\projects\linq2db\Source\Linq\Builder\SelectContext.cs:line 94   at LinqToDB.Linq.Builder.SelectContext.BuildQuery[T](Query`1 query, ParameterExpression queryParameter) in c:\projects\linq2db\Source\Linq\Builder\SelectContext.cs:line 66   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]() in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.cs:line 161   at LinqToDB.Linq.Query`1.GetQuery(IDataContextInfo dataContextInfo, Expression expr) in c:\projects\linq2db\Source\Linq\Query.cs:line 165   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression expression, Boolean cache) in c:\projects\linq2db\Source\Linq\ExpressionQuery.cs:line 89   at LinqToDB.Linq.ExpressionQuery`1.Execute(IDataContextInfo dataContextInfo, Expression expression) in c:\projects\linq2db\Source\Linq\ExpressionQuery.cs:line 76   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.IEnumerable.GetEnumerator() in c:\projects\linq2db\Source\Linq\ExpressionQuery.cs:line 161   at LINQPad.UI.ExplorerGrid.ConvertToList(IEnumerable source)   at LINQPad.UI.ExplorerGrid.UpdateDataSourceToUse()
--
```

### Steps to reproduce
Create table as follows:

```
CREATE TABLE Users (
    UserID int IDENTITY(1,1) NOT NULL,
    SiteID INT NOT NULL,
    Username VARCHAR(256) NOT NULL,
    Active BIT NULL
)
```

Execute following code:

```c#
public partial class TestDB : LinqToDB.Data.DataConnection
{
    public ITable<User> Users { get { return this.GetTable<User>(); } }

    public Context(string connectionString)
        : base("SqlServer", connectionString)
    {
    }
}

[Table(Schema="dbo", Name="Users")]
public partial class User
{
    [Identity]
    public int UserID { get; set; }
    [Column, NotNull]
    public int SiteID { get; set; }
    [Column, NotNull]
    public string Username { get; set; }
    [Column, NotNull]
    public bool? Active { get; set; }
}

public class Program
{
    const string connStr = @"insert your connection string here";
    void Main()
    {
        using (var db = new TestDB(connStr))
        {
            var query = 
                from u in db.Users
                group u.Active ? 1 : 0 by u.SiteID into grp
                select new 
                {
                    SiteID = grp.Key,
                    Total = grp.Count(),
                    Active = grp.Sum(),
                    Inactive = grp.Count(_ => _ == 0),
            var results = query.ToArray();
        }
    }
}
```

On enumeration the exception message should be shown.

Can be worked around by modifying the query to use `Sum(_ => 1 - _)` instead, which generates the correct output.

During testing I found other issues with the `Count(predicate)` handling that seem to indicate that it is broken in interesting ways.

### Environment details
linq2db version:  *1.0.7.6*, *1.10.1.0*, *1.10.2.0*
Database Server: *Microsoft SQL Server 2012*  
Database Provider: 
Operating system: *Windows 10*  
Framework version: *.NET Framework 4.6*

[1]: https://msdn.microsoft.com/en-us/library/bb534807(v=vs.110).aspx

**Comment by sdanyliv** [comment]
@Corey-M, there is special aggregate for such kind of queries.
`CountExt` - you can pass expression that returns null if you plan to filter out record. 
Also it has overloads to support full [`COUNT`](https://docs.microsoft.com/en-us/sql/t-sql/functions/count-transact-sql) syntax.

**Comment by sdanyliv** [comment]
It is a known issue. Tried to fix it by myself, but better to leave fix to @igor-tkachev.

**Comment by MaceWindu** [comment]
I'm adding this issue to 3.5.2 release: looks like it was fixed in 3.5.2 or before (regression tests pass in master)
